File: src/jsonrpc/engine/getPayloadV4/engine_getPayloadV4.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Obtains execution payload from payload build process
///
/// Example:
/// Payload id: "0x0000000038fa5dd"
/// Result: ...
///
/// Implements the `engine_getPayloadV4` JSON-RPC method.
pub const EngineGetPayloadV4 = @This();

/// The JSON-RPC method name
pub const method = "engine_getPayloadV4";

/// Parameters for `engine_getPayloadV4`
pub const Params = struct {
    /// 8 hex encoded bytes
    payload id: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.payload id);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .payload id = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `engine_getPayloadV4`
pub const Result = struct {
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};

File: packages/voltaire-ts/src/jsonrpc/engine/getPayloadV4/engine_getPayloadV4.go
package engine

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Obtains execution payload from payload build process
//
// Example:
// Payload id: "0x0000000038fa5dd"
// Result: ...
//
// Implements the engine_getPayloadV4 JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "engine_getPayloadV4"

// Params represents the parameters for engine_getPayloadV4
type Params struct {
	// 8 hex encoded bytes
	Payload id types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Payload id,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 1 {
		return fmt.Errorf("expected 1 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Payload id); err != nil {
		return fmt.Errorf("parameter 0 (Payload id): %w", err)
	}

	return nil
}

// Result represents the result for engine_getPayloadV4
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}
